### PR TITLE
Allow disabling of types from the command line

### PR DESCRIPTION
**Pull request by Renaud-K**
I would to add hidden options to disable types through the `TargetCharacteristics`. I am seeing issues when I do this programmatically and would like, for anyone, to have the ability to reproduce them for development and testing purposes. 

I am planning to file a couple of issues following this patch.

### CODE REVIEW

File: clang/include/clang/Driver/Options.td [hunk 1]
```diff
@@ -6761,6 +6761,12 @@ def fdefault_integer_8 : Flag<["-"],"fdefault-integer-8">, Group<f_Group>,
   HelpText<"Set the default integer and logical kind to an 8 byte wide type">;
 def fdefault_real_8 : Flag<["-"],"fdefault-real-8">, Group<f_Group>,
   HelpText<"Set the default real kind to an 8 byte wide type">;
+def fdisable_real_3 : Flag<["-"],"fdisable-real-3">, Group<f_Group>,
```
Comment (klausler) [inline clang/include/clang/Driver/Options.td:6764]: maybe also `-fdisable-real-2`?

**Review by klausler — APPROVED** [review]

**Review by banach-space — COMMENTED** [review]
What's the end goal here? To add an option for every possible bit-width? If yes, then these options should take bit width as parameters.

Btw, does GFortran implement anything like this? If yes, are you making sure that the semantics in Flang are consistent? Please also add help text.

**Comment by Renaud-K** [comment]
The end goal is testing for issues that this PR would expose and that I am planning on filing next.
Every bit-width? (5,6,7,9,11,12,13,14,15)  I don't think so.
It does not matter what gfortran does. It is not a feature, it is hidden for testing purposes.

**Review by vzakhari — APPROVED** [review]
Looks good to me, but maybe we can have just a single option that accepts predefined words like `real2`, `integer4`, etc.  I wonder if we can process multiple instances of an option like https://github.com/llvm/llvm-project/blob/0ad6cee926865d7210eed9e67bfb20dce19c6633/clang/include/clang/Driver/Options.td#L5086.  This is just a suggestion to make the coverage of types/kinds more complete.

**Review by tblah — APPROVED** [review]
I agree with other reviewers that a generic option would be better than a list of particular types. But for me that's just a nit because these are hidden debug options anyway

**Comment by Renaud-K** [comment]
I have been looking into this but long story short, but I am little pressed by time and was not planning on spending too much time on this. For the need that we have, it should be ok.

**Pull request merged**